littlepay/api/funding_sources.py

```diff
@@ -0,0 +1,40 @@
+from dataclasses import dataclass
+from typing import List, Optional
+
+from littlepay.api import ClientProtocol
+
+
+@dataclass
+class FundingSourceResponse:
+    id: str
+    card_first_digits: str
+    card_last_digits: str
+    card_expiry_month: str
+    card_expiry_year: str
+    card_scheme: str
+    form_factor: str
+    participant_id: str
+    is_fpan: bool
+    related_funding_sources: List[dict]
```

Comment:
Would it work to call this a `List[FundingSourceResponse]`? Since it seems like this is sort of a recursive definition based on the docs?

Reply:
I agree it'd be nice from a API consumer's perspective for this to be a `List[FundingSourceResponse]`. However, just changing the type hint here doesn't actually do anything to the data, i.e. it doesn't automatically convert each `dict` into a `FundingSourceResponse` for us. 

I tried it out and saw that the type of each list item is still `dict`:
```
> funding_source.related_funding_sources[0]
{'id': '0d1943d5-f145-4a36-9...1465b52350', 'card_first_digits': '411111', 'card_last_digits': '1111', 'card_expiry_month': '09', 'card_expiry_year': '28', 'card_scheme': 'visa', 'form_factor': 'unknown', 'participant_id': 'mst', 'is_fpan': False}
> type(funding_source.related_funding_sources[0])
<class 'dict'>
> funding_source.related_funding_sources[0].id
Traceback (most recent call last):
  File "<string>", line 1, in <module>
AttributeError: 'dict' object has no attribute 'id'
```

We'd need to change our [implementation for object instantiation](https://github.com/cal-itp/littlepay/blob/main/littlepay/api/client.py#L129) to somehow detect these recursive type of fields or implement a `__post_init__` method on the dataclass.

 - https://stackoverflow.com/questions/51199031/python-3-dataclass-initialization
 - https://stackoverflow.com/questions/51737828/python-dataclasses-what-type-to-use-if-post-init-performs-type-conversion

Reply:
Ah, yeah, that all sounds too complicated. Maybe we can revisit later the `__post_init__` idea but not worth worrying about right now.